File: src/http-client-mock/RequestCapture.cs
using System;
 using System.IO;
 using System.Net.Http;
 using System.Net.Http.Json;
using JetBrains.Annotations;

namespace HttpClientMock
{
    public class RequestCapture
    {
        private readonly HttpContent content;
        public RequestCapture(Uri requestUri, HttpMethod method, HttpContent content)
        {
            Method = method;
            this.content = content;
            RequestUri = requestUri;
        }
        
        [PublicAPI]
        public HttpMethod Method { get; }
        [PublicAPI]
        public Uri RequestUri { get; }

        public T Model<T>(T schema = default)
        {
            return content is EmptyContent ? 
                default : Read<T>(Copy(content));
        }
        
        private static T Read<T>(HttpContent content)
        {
            if (typeof(T) == typeof(string))
            {
                return (T) (object) content.ReadAsStringAsync().Result;
            }
            if (typeof(T) == typeof(Stream))
            {
                return (T) (object) content.ReadAsStreamAsync().Result;
            }
            if (typeof(T) == typeof(byte[]))
            {
                return (T) (object) content.ReadAsByteArrayAsync().Result;
            }
            return content.ReadFromJsonAsync<T>().Result;
        }

        private static StreamContent Copy(HttpContent content)
        {
            var memoryStream = new MemoryStream();
            content.CopyToAsync(memoryStream).Wait();
            memoryStream.Position = 0;
            var streamContent = new StreamContent(memoryStream);
            foreach (var (key, value) in content.Headers)
            {
                streamContent.Headers.Add(key, value);    
            }
            return streamContent;
        }
    }
}

File: test/unit-test/HttpClientMockx/RequestCapture.fs
namespace HttpClientMockx

open System;
open System.IO;
open System.Net.Http;
open System.Net.Http.Json;
open JetBrains.Annotations;


type RequestCapture (requestUri: Uri ,  method: HttpMethod,  content: HttpContent option) =
    [<PublicAPI>]
    member this.HttpMethod = method
    
    [<PublicAPI>]
    member this.RequestUri = requestUri

    static member Read<'T>(content: HttpContent): 'T =
        match typeof<'T> with
            | t when t = typeof<string> -> (box (content.ReadAsStringAsync().Result) ) :?> 'T
            | t when t = typeof<Stream> -> (box (content.ReadAsStreamAsync().Result)) :?> 'T
            | t when t = typeof<byte[]> -> (box (content.ReadAsByteArrayAsync().Result)) :?> 'T
            | _ -> content.ReadFromJsonAsync<'T>().Result
            
    static member Copy(content: HttpContent ): StreamContent =
        let memoryStream = new MemoryStream()
        content.CopyToAsync(memoryStream).Wait()
//        async { do! content.CopyToAsync(memoryStream) |> Async.AwaitTask } |> Async.RunSynchronously

        memoryStream.Position <- 0L
        let streamContent = new StreamContent(memoryStream);
        for header in content.Headers do
            streamContent.Headers.Add(header.Key, header.Value)
        streamContent
    
    member this.Model<'T>(): 'T option =
        match content with
            | None -> None
            | Some c -> Some(RequestCapture.Read<'T>(RequestCapture.Copy(c)))
